samza-core/src/main/java/org/apache/samza/zk/ScheduleAfterDebounceTime.java

```diff
@@ -20,99 +20,164 @@
 
-  synchronized public void scheduleAfterDebounceTime(String actionName, long debounceTimeMs, Runnable runnable) {
-    // check if this action has been scheduled already
-    ScheduledFuture sf = futureHandles.get(actionName);
-    if (sf != null && !sf.isDone()) {
-      LOG.info("cancel future for " + actionName);
-      // attempt to cancel
-      if (!sf.cancel(false)) {
+  /**
+   * Stops the scheduler. After this invocation no further schedule calls will be accepted
+   * and all pending enqueued tasks will be cancelled.
+   */
+  public void stopScheduler() {
+    scheduledExecutorService.shutdownNow();
+
+    // Clear the existing future handles.
+    synchronized (lock) {
+      futureHandles.clear();
+    }
+  }
+
+  /**
+   * Tries to cancel the task that belongs to {@code actionName} submitted to the queue.
+   *
+   * @param actionName the name of action to cancel.
+   */
+  private void tryCancelScheduledAction(String actionName) {
+    ScheduledFuture scheduledFuture = futureHandles.get(actionName);
+    if (scheduledFuture != null && !scheduledFuture.isDone()) {
+      LOG.info("Attempting to cancel the future of action: {}", actionName);
+      // Attempt to cancel
+      if (!scheduledFuture.cancel(false)) {
         try {
-          sf.get(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+          scheduledFuture.get(TIMEOUT_MS, TimeUnit.MILLISECONDS);
         } catch (Exception e) {
           // we ignore the exception
-          LOG.warn("cancel for action " + actionName + " failed with ", e);
+          LOG.warn("Cancelling the future of action: {} failed.", new Object[]{actionName, e});
         }
       }
       futureHandles.remove(actionName);
     }
-    // schedule a new task
-    sf = scheduledExecutorService.schedule(() -> {
-        try {
-          runnable.run();
-          LOG.debug(actionName + " completed successfully.");
-        } catch (Throwable t) {
-          LOG.error(actionName + " threw an exception.", t);
-          if (scheduledTaskFailureCallback != null) {
-            scheduledTaskFailureCallback.onError(t);
-          }
+  }
+
+  /**
+   * Decorate the executable action with exception handlers to facilitate cleanup on failures.
+   *
+   * @param actionName the name of the scheduleable action.
+   * @param runnable the action to execute.
+   * @return the executable action decorated with exception handlers.
+   */
+  private Runnable getScheduleableAction(String actionName, Runnable runnable) {
+    return () -> {
+      try {
+        runnable.run();
+        LOG.debug("Action: {} completed successfully.", actionName);
+        /*
+         * Expects all run() implementations <b>not to swallow the interrupts.</b>
+         * This thread is interrupted from an external source(mostly executor service) to die.
+         */
+        if (Thread.currentThread().isInterrupted()) {
+          LOG.warn("Action: {} is interrupted.", actionName);
+          doCleanUpOnTaskException(new InterruptedException());
```

Comment:
minor suggestion: It is perhaps simpler to set a local variable for storing the exception value. Inspect its value in `finally` clause and then, take action.

Reply:
+1

Reply:
@navina  @sborya 
I also wanted to do that, we need to declare that exception variable as effectively final since this happens in a lambda(either as an array or a Wrapper, which i think is a kill).
I tried adding it that way initially and it looks weird. 
Please take a look at the test (for same thing that i'm doing there)

Reply:
I agree/tried initially to follow the way you are recommending. As we'd discussed offline, declaring it as array or reference (just to make it effectively final) is the reason which make it more convoluted. This way is simplar.